This makes bind_star_handling even more RFC-compliant: If *.example.com. 
and bar.example.com. exist, then foo.bar.example.com. should point to 
"not there" instead of whatever *.example.com</tt> points at.

--- maradns-1.2.02/server/MaraDNS.c	2006-01-12 09:51:12.000000000 -0800
+++ maradns-1.2.03/server/MaraDNS.c	2006-01-12 10:01:13.000000000 -0800
@@ -3337,12 +3337,13 @@
 
     /* Remmy's issue: If we have any recordtype for a given record, we don't
        look for a star record with the same name.  This emulates BIND's
-       behavior for star records */
+       behavior for star records and strictly follows RFC1034 section
+       4.3.3 */
     if(bind_star_handling == 1) {
         if(js_copy(origq,lookfor) == JS_ERROR)
             goto serv_fail;
         if(fold_case(lookfor) == JS_ERROR)
-	    goto serv_fail;
+            goto serv_fail;
         /* Look for an ANY record with the same name and
            goto skip_starwhitis if found */
 	if(change_rtype(lookfor,RR_ANY) == JS_ERROR) {
@@ -3385,6 +3386,24 @@
         spot_data = mhash_get(bighash,lookfor);
         point = spot_data.value;
         if(spot_data.value != 0 && spot_data.datatype == MARA_DNSRR) {
+
+            /* Optional strict RFC1034 section 4.3.3 compliance */
+            if(bind_star_handling == 1) {
+                int rtype_saver;
+                rtype_saver = get_rtype(lookfor);
+	        if(change_rtype(lookfor,RR_ANY) == JS_ERROR) {
+		    goto serv_fail;
+                }
+                spot_data = mhash_get(bighash,lookfor);
+                if(spot_data.value != 0 && 
+                     spot_data.datatype == MARA_DNS_LIST) {
+                    goto skip_starwhitis;
+                }
+                if(change_rtype(lookfor,rtype_saver) == JS_ERROR) {
+                    goto serv_fail;
+                }
+            }
+
             udpstar(spot_data.value,header.id,sock,0,origq,origq,
 			    desires_recursion,0,ect);
             js_destroy(lookfor); js_destroy(origq); js_destroy(lc);
@@ -3431,6 +3450,24 @@
     /* Look for a star record "above" when it is a CNAME */
     while(bobbit_starlabel(lookfor) > 0) {
         spot_data = mhash_get(bighash,lookfor);
+
+        /* Optional strict RFC1034 section 4.3.3 compliance */
+        if(bind_star_handling == 1) {
+            int rtype_saver;
+            rtype_saver = get_rtype(lookfor);
+            if(change_rtype(lookfor,RR_ANY) == JS_ERROR) {
+                goto serv_fail;
+            }
+            spot_data = mhash_get(bighash,lookfor);
+            if(spot_data.value != 0 && 
+                 spot_data.datatype == MARA_DNS_LIST) {
+                goto skip_starwhitis;
+            }
+            if(change_rtype(lookfor,rtype_saver) == JS_ERROR) {
+                goto serv_fail;
+            }
+        }
+
         if(spot_data.value != 0 && spot_data.datatype == MARA_DNSRR) {
             if(js_copy(origq,lookfor) == JS_ERROR)
                 goto serv_fail;
